pkg/collector/container.go

```diff
@@ -129,7 +143,108 @@ func getContainerPorts(logger logr.Logger, cfg string) map[string]corev1.Contain
 	return ports
 }
 
-func portMapToList(portMap map[string]corev1.ContainerPort) []corev1.ContainerPort {
+func getServicePortsFromCWAgentConfig(logger logr.Logger, configJSON *adapters.CwaConfig) []corev1.ServicePort {
+	servicePortsMap := getAppSignalsServicePortsMap()
+	getMetricsReceiversServicePorts(logger, configJSON, servicePortsMap)
+	getLogsReceiversServicePorts(logger, configJSON, servicePortsMap)
+	getTracesReceiversServicePorts(logger, configJSON, servicePortsMap)
+	return PortMapToServicePortList(servicePortsMap)
+}
+
+func getMetricsReceiversServicePorts(logger logr.Logger, configJSON *adapters.CwaConfig, containerPortsMap map[int32]corev1.ServicePort) {
+	if configJSON.Metrics == nil || configJSON.Metrics.MetricsCollected == nil {
+		return
+	}
+	//StatD - https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-custom-metrics-statsd.html
+	if configJSON.Metrics.MetricsCollected.StatsD != nil {
+		getReceiverServicePort(logger, configJSON.Metrics.MetricsCollected.StatsD.ServiceAddress, StatsD, corev1.ProtocolUDP, containerPortsMap)
+	}
+	//CollectD - https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-custom-metrics-collectd.html
+	if configJSON.Metrics.MetricsCollected.CollectD != nil {
+		getReceiverServicePort(logger, configJSON.Metrics.MetricsCollected.CollectD.ServiceAddress, CollectD, corev1.ProtocolUDP, containerPortsMap)
+	}
+}
+
+func getReceiverServicePort(logger logr.Logger, serviceAddress string, receiverName string, protocol corev1.Protocol, containerPortsMap map[int32]corev1.ServicePort) {
```

Comment:
it's not a good practice to modify the input parameters. Can we just return the ServicePort object and then the caller will modify the containerPortsMap?

Reply:
+1

Reply:
In Go, it is idiomatic to modify structures passed in as parameters rather than returning newly created ones, especially for larger data structures. This is because Go uses pointers to pass data structures, and modifying the structure in place can be more efficient than creating a new one and returning it.